### PR TITLE
Reference types changes to remove subtyping

### PR DESCRIPTION
**Pull request by binji**
Main changes:
* Rename `anyref` -> `externref`
* Remove `nullref`
* Rename `hostref` -> `externref`
* `ref.null` and `ref.is_null` now have "ref kind" parameter
* Add ref kind keywords: `func`, `extern`, `exn`

### CODE REVIEW

**Comment by binji** [comment]
This PR can't land as-is, since it fails all the spec tests (that need to be updated)

File: src/lexer-keywords.txt [hunk 1]
```diff
@@ -15,8 +17,7 @@ struct TokenInfo {
   };
 };
 %%
-anyref, Type::Anyref
-array, TokenType::Array
+array, Type::Array, TokenType::Array
```
[the remark this answers is not in the document]
Comment (sbc100) [reply src/lexer-keywords.txt:20]: Maybe another possible split out PR?

**Comment by sbc100** [comment]
Is there a PR in the reftypes proposal you can link to here?

**Comment by binji** [comment]
Here's the link to the reference-types PR: https://github.com/WebAssembly/reference-types/pull/87. I'll add it to the commit too.

File: src/lexer-keywords.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (binji) [reply src/lexer-keywords.txt:20]: Done.

File: src/interp/interp-inl.h [hunk 2]
```diff
@@ -365,16 +365,7 @@ template <typename T> void RequireType(ValueType type) {
...
-    return true;
-  }
-  if (!IsReference(expected)) {
-    return false;
-  }
-  if (expected == ValueType::Anyref || actual == ValueType::Nullref) {
-    return true;
-  }
-  return false;
+  return expected == actual;
```
[the remark this answers is not in the document]
Comment (binji) [reply src/interp/interp-inl.h:370]: Done.

File: src/interp/interp-util.cc [hunk 3]
```diff
@@ -49,21 +49,15 @@ std::string TypedValueToString(const TypedValue& tv) {
...
-    case Type::Nullref:
-      return StringPrintf("nullref");
-
-    case Type::Hostref:
-      return StringPrintf("hostref:%" PRIzd, tv.value.Get<Ref>().index);
-
     case Type::Funcref:
       return StringPrintf("funcref:%" PRIzd, tv.value.Get<Ref>().index);
 
+    case Type::Externref:
```
[the remark this answers is not in the document]
Comment (binji) [reply src/interp/interp-util.cc]: Done.

**Pull request merged**